### PR TITLE
[stableswap]: simplify CFMM solver 

### PR DESCRIPTION
**Pull request by AlpinYukseloglu**
Closes: #2696 

## What is the purpose of the change

This PR implements a drastically simplified solver for our two-asset stableswap CFMM.

Key differences:
- Fits on a single screen (core solver logic is ~5 lines)
- _Far_ fewer square root, product, and power terms
- Largest power is same as previous (8)
- No negative numbers in intermediate calculations
- Much easier to parse visually

## Brief Changelog

I derived this solver manually, but we can confirm on Desmos that it fits our curve exactly: https://www.desmos.com/calculator/wxl9ffxofw

As described in the issue, the key simplifcations leaned on the following assumptions:
- For any given swap, everything except for the amount of the output asset is known & constant
- The number of assets in the pool are always greater than zero (let's us use the original CFMM as part of our solver)
- Finding the final asset reserve amounts is as good as finding the output amount since we can subtract the original reserves at the end

Pseudocode for new solver:
```
func solveCfmm(xReserve, yReserve, yIn) {
    Calculate k using xReserve and yReserve

    yReserve = yReserve + yIn
    
    Use new yReserve to solve for new xReserve (using new solver derived from xy(x^2 + y^2) = k)

    xOut = original xReserve minus resulting output

    Return xOut
}
```

## Testing and Verifying

Basic tests pass, and the ones that failed on our original solver are around 3x closer to the target with this solver (likely due to fewer terms leading to fewer approximations).

Still working on more comprehensive tests, although we might first need to solve general precision issues around our CFMM.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

**Comment by ValarDragon** [comment]
Ah great idea, there was no need to keep `b` as a variable in the prior equations!

File: x/gamm/pool-models/stableswap/amm.go [hunk 1]
```diff
@@ -256,9 +272,59 @@ func solveCfmmMulti(xReserve, yReserve, wSumSquares, yIn sdk.Dec) sdk.Dec {
...
+	}
+
 	return a
 }
 
+// solidly CFMM is xy(x^2 + y^2) = k
+// So we want to solve for a given addition of `b` units of y into the pool,
+// how many units `a` of x do we get out.
+// So we solve the following expression for `a`
+// xy(x^2 + y^2) = (x - a)(y + b)((x - a)^2 + (y + b)^2)
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/amm.go:286]:
```suggestion
// Let y' = y + b
// we solve k = (x'y')(x'^2 + y^2) for x', using the following equation {wolfram alpha link}
// Then we use that to derive the change in x as x_out = x' - x
```

**Comment by AlpinYukseloglu** [comment]
Recent commit cuts the exponents of the two largest terms in half (`x^8 -> x^4` and `k^2 -> k`)

This makes meaningful progress towards #2701, but we still overflow at much smaller asset amounts than we expect to be able to accommodate (current capacity is just under $10m per asset in a 2-asset pool, and we need to be able to support $100m per asset in a 3+ asset pool)

Reference to verify new implementation: https://www.desmos.com/calculator/bx5m5wpind

The last idea/trick I have in mind is potentially dropping the +1s given the size of pools we're dealing with (maybe have two solvers and switch to the other one when x, y, or k is sufficiently large)

Ultimately we'll want to use our binary search solver, but I believe it's important to have a comprehensive and simple reference solver to benchmark against in the meantime

**Comment by ValarDragon** [comment]
nice, this is great! Big win!

(The largest term is x^7 still tho =p, `k = O(x^4)`)

Heres an equation switch, basically moving `(x^2k)^{1/3}` to instead be `x (k/x)^{1/3}`, which then truly lowers the largest term to x^4 as far as I can tell!

And actually if you unwrap `k` in those equations, by making `k' = k/x = y(x^2 + y^2)`, and computing off of `k'` directly everywhere, you'd actually be at `O(x^3)` everywhere AFAICT!

**Comment by AlpinYukseloglu** [comment]
This is awesome! To your second point: can't we just exponentiate wherever we see a `k / x` to take it from `O(x^4)` to `O(x^3)` without ever needing to unwrap the k?

e.g. `k / x = (sqrt(k) / sqrt(x))^2` so we reach `O(x^3)` without any intermediate terms > `O(x^2)`?

**Comment by ValarDragon** [comment]
That looks great to me!

File: x/gamm/pool-models/stableswap/amm.go [hunk 2]
```diff
@@ -276,6 +281,59 @@ func solveCfmmMulti(xReserve, yReserve, wSumSquares, yIn osmomath.BigDec) osmoma
 	return a
 }
 
+// solidly CFMM is xy(x^2 + y^2) = k
+// So we want to solve for a given addition of `b` units of y into the pool,
+// how many units `a` of x do we get out.
+// Let y' = y + b
+// we solve k = (x'y')(x'^2 + y^2) for x', using the following equation {wolfram alpha link}
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/amm.go:288]: Needs wolfram alpha link, or equation we test and explanation of correctness

File: x/gamm/pool-models/stableswap/amm.go [hunk 3]
```diff
@@ -276,6 +281,59 @@ func solveCfmmMulti(xReserve, yReserve, wSumSquares, yIn osmomath.BigDec) osmoma
...
+	y3 := y2.Mul(y_new)
+	y4 := y3.Mul(y_new)
+	large_term := (y4.Quo(k)).Mul(osmomath.NewBigDec(2).Quo(twentySevenRootTwo))
+	large_term2 := large_term.Mul(large_term)
+
+	// solve for new xReserve using new yReserve and old k using a solver derived from xy(x^2 + y^2) = k
+	sqrt_term, err := (osmomath.OneDec().Add(large_term2)).ApproxRoot(2)
+	if err != nil {
+		panic(err)
+	}
```
Comment (ValarDragon) [inline x/gamm/pool-models/stableswap/amm.go:316]: Lets take the equation we want to compute, write out the variable substitutions we do, then show that each of this is very clearly computing the desired variable

Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/amm.go]: Took a crack at this in the recent commit!

File: x/gamm/pool-models/stableswap/amm.go (hunk 2 above)
Comment (AlpinYukseloglu) [reply x/gamm/pool-models/stableswap/amm.go]: I put a wolfram alpha link of the original equation, a desmos link for our simplified one, and direct equation + our key abstractions in my last commit

**Review by ValarDragon — APPROVED** [review]
Nice!

**Pull request merged**